pallets/farming/src/lib.rs

```diff
@@ -0,0 +1,839 @@
+        #[pallet::call_index(0)]
+        #[pallet::weight(T::WeightInfo::create_farming_pool())]
+        pub fn create_farming_pool(
+            origin: OriginFor<T>,
+            tokens_proportion: Vec<(CurrencyIdOf<T>, Perbill)>,
+            basic_rewards: Vec<(CurrencyIdOf<T>, BalanceOf<T>)>,
+            gauge_init: Option<GaugeInitType<T>>,
+            min_deposit_to_start: BalanceOf<T>,
+            #[pallet::compact] after_block_to_start: BlockNumberFor<T>,
+            #[pallet::compact] withdraw_limit_time: BlockNumberFor<T>,
+            #[pallet::compact] claim_limit_time: BlockNumberFor<T>,
+            withdraw_limit_count: u8,
+        ) -> DispatchResult {
+            T::ControlOrigin::ensure_origin(origin)?;
+
+            let pid = Self::pool_next_id();
+            let keeper = T::Keeper::get().into_sub_account_truncating(pid);
+            let reward_issuer = T::RewardIssuer::get().into_sub_account_truncating(pid);
+            let basic_token = tokens_proportion[0];
+            let tokens_proportion_map: BTreeMap<CurrencyIdOf<T>, Perbill> =
+                tokens_proportion.into_iter().map(|(k, v)| (k, v)).collect();
+            let basic_rewards_map: BTreeMap<CurrencyIdOf<T>, BalanceOf<T>> =
+                basic_rewards.into_iter().map(|(k, v)| (k, v)).collect();
+
+            let mut pool_info = PoolInfo::new(
+                keeper,
+                reward_issuer,
+                tokens_proportion_map,
+                basic_token,
+                basic_rewards_map,
+                None,
+                min_deposit_to_start,
+                after_block_to_start,
+                withdraw_limit_time,
+                claim_limit_time,
+                withdraw_limit_count,
+            );
+
+            if let Some((gauge_token, max_block, gauge_basic_rewards)) = gauge_init {
+                let gauge_basic_rewards_map: BTreeMap<CurrencyIdOf<T>, BalanceOf<T>> =
+                    gauge_basic_rewards
+                        .into_iter()
+                        .map(|(k, v)| (k, v))
+                        .collect();
+
+                Self::create_gauge_pool(
+                    pid,
+                    &mut pool_info,
+                    gauge_token,
+                    gauge_basic_rewards_map,
+                    max_block,
+                )?;
+            };
+
+            PoolInfos::<T>::insert(pid, &pool_info);
+            PoolNextId::<T>::mutate(|id| -> DispatchResult {
+                *id = id.checked_add(1).ok_or(ArithmeticError::Overflow)?;
+                Ok(())
+            })?;
+
+            Self::deposit_event(Event::FarmingPoolCreated { pid });
+            Ok(())
+        }
+
+        /// After create farming pool, need to deposit reward token into the pool.
+        #[pallet::call_index(1)]
+        #[pallet::weight(0)]
+        pub fn charge(
+            origin: OriginFor<T>,
+            pid: PoolId,
+            rewards: Vec<(CurrencyIdOf<T>, BalanceOf<T>)>,
```

Comment:
Some one might split reward tokens into like 1000 rewards here, like
```
[
    (1, 1),
    (1, 1),
    (1, 1),
    ...
]
```

Reply:
same above, if someone want to send token to us, it's fine.

Reply:
could be used for ddos ?

Reply:
add weight to `Charge` should fix that.